src/Traverser.h

```diff
@@ -358,107 +346,93 @@ class DualTraverser : public Traverser<Data> {
           for (int j = 0; j < node->n_children; j++) {
             nodes.push(node->getChild(j));
           }
         } else {
-          Visitor::node(*node, *source_leaf);
+          doNode<Visitor>(source_leaf, node, tp.r_local);
         }
       }
     }
   }
 
-  virtual void resumeTrav(Key new_key) override {
-    auto& now_ready = curr_nodes[new_key];
-    std::vector<std::pair<Key, Node<Data>*>> curr_nodes_insertions;
-    Node<Data>* start_node = nullptr;
-    auto && resume_nodes = tp.r_local->resume_nodes_per_tp[tp.thisIndex];
-    if (!resume_nodes.empty()) {
-      start_node = resume_nodes.front();
-      CkAssert(start_node->key == new_key);
+  virtual void resumeTrav() override {
+    auto && resume_nodes = tp.r_local->resume_nodes_per_part[tp.thisIndex];
+    while (!resume_nodes.empty()) {
+      doTrav(resume_nodes.front());
       resume_nodes.pop();
     }
-    else {
-      start_node = tp.global_root->getDescendant(new_key);
-      if (start_node == nullptr) CkAbort("If doing dual traversal, don't set num_share_nodes");
-    }
-    for (auto new_payload : now_ready) {
-      std::stack<std::pair<Node<Data>*, Node<Data>*>> nodes;
-      nodes.push(std::make_pair(start_node, new_payload));
-      while (nodes.size()) {
-        Node<Data>* node = nodes.top().first, *curr_payload = nodes.top().second;
-        // node is target, payload is source
-        nodes.pop();
+  }
+
+  void doTrav(Node<Data>* start_node) {
+    Key new_key = start_node->key;
+    std::vector<std::pair<Key, Node<Data>*>> curr_nodes_insertions;
+    auto& now_ready = curr_nodes[new_key];
+    std::stack<std::pair<Node<Data>*, Node<Data>*>> nodes;
+    for (auto && np : now_ready) nodes.emplace(start_node, np);
+    while (!nodes.empty()) {
+      Node<Data>* node = nodes.top().first, *curr_payload = nodes.top().second;
+      // node is source, payload is target
+      nodes.pop();
 #if DEBUG
-        CkPrintf("tp %d, key = %d, type = %d, pe %d\n", tp.thisIndex, node->key, node->type, CkMyPe());
+      CkPrintf("tp %d, target key = %d, type = %d, source key = %d, type = %d, pe %d\n", tp.thisIndex, node->key, node->type, curr_payload->key, curr_payload->type, CkMyPe());
 #endif
-        if (curr_payload->type == Node<Data>::Type::EmptyLeaf) {
-          continue;
-        }
-        else if (curr_payload->type == Node<Data>::Type::Leaf) {
-          int leaf_index = -1;
-          for (int i = 0; i < tp.leaves.size(); i++) {
-            if (tp.leaves[i] == curr_payload) {
-              leaf_index = i;
-              break;
-            }
+      if (curr_payload->type == Node<Data>::Type::EmptyLeaf) {
+        continue;
+      }
+      CkAssert(curr_payload->type == Node<Data>::Type::Internal);
+      switch (node->type) {
+        case Node<Data>::Type::Leaf:
+        case Node<Data>::Type::CachedRemoteLeaf:
+          {
+            doLeaf<Visitor>(node, curr_payload, tp.r_local); // n2 calc
```

Comment:
@trquinn this okay? this is when we've arrived on a source node thats a leaf but our target payload is not a leaf.

Reply:
if not, we can call runInvertedTraversal from there.